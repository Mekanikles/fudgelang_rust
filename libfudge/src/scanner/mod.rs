pub mod token;
pub use token::*;
pub mod scanner;
pub use scanner::*;

#[cfg(test)]
mod test;
