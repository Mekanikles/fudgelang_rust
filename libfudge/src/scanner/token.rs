use crate::source;
use std::fmt;
use std::str;

#[derive(PartialEq, Debug, Copy, Clone)]
pub enum TokenType {
    // 1-char tokens
    Comma,
    Dot,
    Colon,
    SemiColon,
    Equals,
    Hash,
    OpeningParenthesis,
    ClosingParenthesis,
    OpeningSquareBracket,
    ClosingSquareBracket,
    OpeningCurlyBrace,
    ClosingCurlyBrace,
    Arrow,
    FatArrow,
    Plus,
    Minus,
    Slash,
    Star,
    GreaterThan,
    LessThan,

    // n-char tokens
    LineBreak,
    Indentation,
    Comment,
    CompareEq,
    GreaterThanOrEq,
    LessThanOrEq,

    // Keywords
    Module,
    If,
    Then,
    Else,
    ElseIf,
    True,
    False,
    Def,
    Var,
    Func,
    Do,
    End,
    Return,

    // Tokens with significant data
    Identifier,
    StringLiteral,
    CharacterLiteral,
    NumericLiteral,
}

#[derive(PartialEq, Debug, Clone, Copy)]
pub struct Token {
    pub tokentype: TokenType,
    pub source_span: source::SourceSpan,
}

impl Token {
    pub fn new(tokentype: TokenType, pos: u64, len: usize) -> Token {
        Token {
            tokentype: tokentype,
            source_span: source::SourceSpan { pos, len },
        }
    }
}

pub struct TokenDisplay<'a> {
    pub token: &'a Token,
    pub source: &'a source::Source,
}

// Debug formatter
impl<'a> fmt::Debug for TokenDisplay<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.token.tokentype {
            TokenType::Identifier
            | TokenType::NumericLiteral
            | TokenType::StringLiteral
            | TokenType::CharacterLiteral => {
                self.token.tokentype.fmt(f).unwrap();
                f.debug_tuple("")
                    .field(&self.token.source_span.pos)
                    .field(&self.token.source_span.len)
                    .field(&str::from_utf8(&self.source.get_span(&self.token.source_span)).unwrap())
                    .finish()
            }
            _ => {
                self.token.tokentype.fmt(f).unwrap();
                f.debug_tuple("")
                    .field(&self.token.source_span.pos)
                    .field(&self.token.source_span.len)
                    .finish()
            }
        }
    }
}
