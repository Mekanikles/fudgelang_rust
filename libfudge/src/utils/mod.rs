pub mod stringstore;

use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[derive(Debug, Clone)]
pub struct StringKey {
    pub key: u64,
    #[cfg(debug_assertions)]
    debugname: String,
}

impl StringKey {
    fn create_key(string: &str) -> u64 {
        let mut hasher = DefaultHasher::new();
        string.hash(&mut hasher);
        return hasher.finish();
    }

    #[cfg(debug_assertions)]
    pub fn from_str(string: &str) -> StringKey {
        StringKey {
            key: Self::create_key(string),
            debugname: string.to_string(),
        }
    }

    #[cfg(not(debug_assertions))]
    pub fn from_str(string: &str) -> StringKey {
        StringKey {
            key: Self::create_key(string),
        }
    }

    #[cfg(debug_assertions)]
    pub fn from_hash(hash: &u64) -> StringKey {
        StringKey {
            key: *hash,
            debugname: "<Raw Hash>".into(),
        }
    }

    #[cfg(not(debug_assertions))]
    pub fn from_hash(hash: &u64) -> StringKey {
        StringKey { key: hash }
    }
}

impl Hash for StringKey {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.key.hash(state);
    }
}

impl PartialEq for StringKey {
    fn eq(&self, other: &Self) -> bool {
        self.key == other.key
    }
}
impl Eq for StringKey {}

impl std::fmt::Display for StringKey {
    #[cfg(debug_assertions)]
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}", self.debugname)
    }

    #[cfg(not(debug_assertions))]
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}", self.key)
    }
}
