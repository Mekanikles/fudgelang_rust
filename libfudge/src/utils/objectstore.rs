use std::collections::HashMap;
use std::hash::Hash;

pub trait ObjectStore<KeyT, ObjectT> {
    type Key;

    fn add(&mut self, object: ObjectT) -> KeyT;
    fn has(&self, key: &KeyT) -> bool;
    fn get(&self, key: &KeyT) -> &ObjectT;
    fn get_mut(&mut self, key: &KeyT) -> &mut ObjectT;
    fn try_get(&self, key: &KeyT) -> Option<&ObjectT>;
    fn try_get_mut(&mut self, key: &KeyT) -> Option<&mut ObjectT>;
}

#[derive(Debug)]
pub struct IndexedObjectStore<ObjectT> {
    objects: Vec<ObjectT>,
}

impl<ObjectT> IndexedObjectStore<ObjectT> {
    pub fn new() -> Self {
        Self {
            objects: Vec::new(),
        }
    }

    pub fn keys(&self) -> std::ops::Range<usize> {
        0..self.objects.len()
    }

    pub fn values(&self) -> &Vec<ObjectT> {
        &self.objects
    }

    pub fn values_mut(&mut self) -> &mut Vec<ObjectT> {
        &mut self.objects
    }

    pub fn remove_vec(&mut self) -> Vec<ObjectT> {
        std::mem::replace(&mut self.objects, Vec::new())
    }

    pub fn into_iter(&mut self) -> <Vec<ObjectT> as IntoIterator>::IntoIter {
        let objects = std::mem::replace(&mut self.objects, Vec::new());
        objects.into_iter()
    }
}

impl<ObjectT> ObjectStore<usize, ObjectT> for IndexedObjectStore<ObjectT> {
    type Key = usize;

    fn add(&mut self, object: ObjectT) -> Self::Key {
        self.objects.push(object);
        self.objects.len() - 1
    }
    fn has(&self, key: &Self::Key) -> bool {
        self.objects.len() > *key
    }
    fn get(&self, key: &Self::Key) -> &ObjectT {
        &self.objects[*key]
    }
    fn get_mut(&mut self, key: &Self::Key) -> &mut ObjectT {
        self.objects.get_mut(*key).unwrap()
    }
    fn try_get(&self, key: &Self::Key) -> Option<&ObjectT> {
        self.objects.get(*key)
    }
    fn try_get_mut(&mut self, key: &Self::Key) -> Option<&mut ObjectT> {
        self.objects.get_mut(*key)
    }
}

pub trait HashedStoreKey<ObjectT>: Hash + Eq + Clone {
    fn from_obj(object: &ObjectT) -> Self;
}

#[derive(Debug)]
pub struct HashedObjectStore<KeyT: HashedStoreKey<ObjectT>, ObjectT: Eq> {
    objects: HashMap<KeyT, ObjectT>,
}

impl<KeyT: HashedStoreKey<ObjectT>, ObjectT: Eq> HashedObjectStore<KeyT, ObjectT> {
    pub fn new() -> Self {
        Self {
            objects: HashMap::new(),
        }
    }

    pub fn keys(&self) -> std::collections::hash_map::Keys<'_, KeyT, ObjectT> {
        self.objects.keys()
    }

    pub fn values(&self) -> std::collections::hash_map::Values<'_, KeyT, ObjectT> {
        self.objects.values()
    }
}

impl<KeyT: HashedStoreKey<ObjectT>, ObjectT: Eq> ObjectStore<KeyT, ObjectT>
    for HashedObjectStore<KeyT, ObjectT>
{
    type Key = KeyT;

    fn add(&mut self, object: ObjectT) -> KeyT {
        let key = KeyT::from_obj(&object);

        if self.objects.contains_key(&key) {
            assert!(object == *self.get(&key));
            return key;
        }

        self.objects.insert(key.clone(), object);

        return key;
    }
    fn has(&self, key: &KeyT) -> bool {
        self.objects.contains_key(key)
    }
    fn get(&self, key: &KeyT) -> &ObjectT {
        &self.objects.get(key).unwrap()
    }
    fn get_mut(&mut self, key: &KeyT) -> &mut ObjectT {
        self.objects.get_mut(key).unwrap()
    }
    fn try_get(&self, key: &KeyT) -> Option<&ObjectT> {
        self.objects.get(key)
    }
    fn try_get_mut(&mut self, key: &KeyT) -> Option<&mut ObjectT> {
        self.objects.get_mut(key)
    }
}
