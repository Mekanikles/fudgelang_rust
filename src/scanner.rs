use crate::source;
use crate::source::SourceReader;

pub struct Scanner<S : source::Source> {
    source : S,
}

impl<S : source::Source> Scanner<S> {
    pub fn new(source : S) -> Self {
        return Scanner::<S> { source };
    }

    pub fn test(&self) {
        let mut reader = self.source.get_reader();
        while let Some(n) = reader.peek() {
            println!("Pos: {}, val: {}", reader.pos(), n);
            reader.advance();
        }
    }
}

