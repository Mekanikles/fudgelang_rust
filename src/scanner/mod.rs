mod token;
mod scanner; pub use scanner::*;

#[cfg(test)]
mod test;
