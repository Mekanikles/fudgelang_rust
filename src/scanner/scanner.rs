use super::*;
use crate::error;
use crate::source;
use phf::phf_map;
use std::debug_assert;
use std::io::Read;
use std::io::Seek;
use std::io::SeekFrom;

use std::io::BufRead;
use std::io::BufReader;

// Map with all scannable keywords
static KEYWORDS: phf::Map<&'static str, TokenType> = phf_map! {
    "if" => TokenType::If,
    "then" => TokenType::Then,
    "else" => TokenType::Else,
    "true" => TokenType::True,
    "false" => TokenType::False,
    "def" => TokenType::Def,
    "func" => TokenType::Func,
    "do" => TokenType::Do,
    "end" => TokenType::End,
    "return" => TokenType::Return,
};

// TODO: This is not a good place for this, has nothing to do with the scanner
pub struct LineInfo {
    pub text: String,
    pub line_start: usize,
    pub row: u32,
}

pub trait Scanner {
    fn get_errors(&self) -> &Vec<error::Error>;
    fn get_token_source_string(&self, token: &Token) -> String;
    fn get_line_info(&self, filepos: u64) -> Option<LineInfo>;
    fn read_token(&mut self) -> Option<Token>;
}

pub struct ScannerImpl<'a, R: Read, S: source::Source<'a, R>> {
    source: &'a S,
    reader: source::LookAheadReader<R>,
    allow_indentation: bool,
    errors: error::ErrorManager,
}

impl<'a, R: Read + Seek, S: source::Source<'a, R>> Scanner for ScannerImpl<'a, R, S> {
    fn get_errors(&self) -> &Vec<error::Error> {
        return self.errors.get_errors();
    }

    fn get_token_source_string(&self, token: &Token) -> String {
        return self.get_source_string(token.source_span.pos, token.source_span.len);
    }

    fn get_line_info(&self, filepos: u64) -> Option<LineInfo> {
        let mut reader = BufReader::new(self.source.get_readable());

        let mut seekpos = 0;
        let mut row = 0;
        let mut text = String::new();
        while let Ok(bytes_read) = reader.read_line(&mut text) {
            if bytes_read == 0 {
                return None;
            }
            let eol = seekpos + bytes_read;
            if eol > filepos as usize {
                return Some(LineInfo {
                    text,
                    line_start: seekpos,
                    row: row + 1,
                });
            }
            seekpos = eol;
            row += 1;
            text.clear();
        }
        return None;
    }

    fn read_token(&mut self) -> Option<Token> {
        let mut invalid_sequence_started = false;
        while let Some(n) = self.reader.peek() {
            // If we have reached the maximum allowed errors before producing a token
            //  consider the scan ended
            if self.errors.reached_error_limit() {
                return None;
            }

            // Indentation management
            if n == b'\t' {
                let indentation_token = self.produce_indentation();
                if !self.allow_indentation {
                    self.errors.log_error(error::new_invalid_indentation_error(
                        indentation_token.source_span.pos,
                        indentation_token.source_span.len as u64,
                    ));
                    continue;
                } else {
                    return Some(indentation_token);
                }
            } else if self.allow_indentation {
                self.allow_indentation = false;
            }

            // Consume unimportant whitespace
            if n == b' ' {
                self.consume_spacing();
                continue;
            }

            // Produce token
            match n {
                b'.' => return Some(self.produce_token_and_advance(TokenType::Dot)),
                b',' => return Some(self.produce_token_and_advance(TokenType::Comma)),
                b':' => return Some(self.produce_token_and_advance(TokenType::Colon)),
                b'=' => return Some(self.produce_token_and_advance(TokenType::Equals)),
                b'+' => return Some(self.produce_token_and_advance(TokenType::Plus)),
                b'-' => match self.reader.lookahead() {
                    Some(b'>') => {
                        return Some(self.produce_token_and_advance_n(TokenType::Arrow, 2))
                    }
                    _ => return Some(self.produce_token_and_advance(TokenType::Minus)),
                },
                b'(' => return Some(self.produce_token_and_advance(TokenType::OpeningParenthesis)),
                b')' => return Some(self.produce_token_and_advance(TokenType::ClosingParenthesis)),
                b'[' => {
                    return Some(self.produce_token_and_advance(TokenType::OpeningSquareBracket))
                }
                b']' => {
                    return Some(self.produce_token_and_advance(TokenType::ClosingSquareBracket))
                }
                b'{' => return Some(self.produce_token_and_advance(TokenType::OpeningCurlyBrace)),
                b'}' => return Some(self.produce_token_and_advance(TokenType::ClosingCurlyBrace)),
                b';' => return Some(self.produce_token_and_advance(TokenType::SemiColon)),
                b'#' => return Some(self.produce_token_and_advance(TokenType::Hash)),
                b'/' => match self.reader.lookahead() {
                    Some(b'/') => return Some(self.produce_linecomment()),
                    Some(b'*') => return Some(self.produce_blockcomment()),
                    _ => return Some(self.produce_token_and_advance(TokenType::Slash)),
                },
                b'*' => match self.reader.lookahead() {
                    Some(b'/') => {
                        self.errors.log_error(error::new_unexpected_sequence_error(
                            self.reader.pos(),
                            2,
                            "Found stray block comment end".into(),
                        ));
                        self.reader.advance();
                        self.reader.advance();
                        continue;
                    }
                    _ => return Some(self.produce_token_and_advance(TokenType::Star)),
                },
                b'\n' => {
                    // TODO: Eat windows-style line breaks as one token
                    self.allow_indentation = true;
                    return Some(self.produce_token_and_advance(TokenType::LineBreak));
                }
                b'\"' => return Some(self.produce_stringliteral()),
                b'\'' => return Some(self.produce_characterliteral()),
                b'0'..=b'9' => return Some(self.produce_numericliteral()),
                b'a'..=b'z' | b'A'..=b'Z' | b'_' => return Some(self.produce_identifier()),
                _ => (),
            }

            // If we reach this point, we did not know what to do with this char
            //  do error handling best we can
            let pos = self.reader.pos();
            if let Some(c) = self.read_utf8_char_with_error() {
                // If this is the start of an "invalid" alphabetic utf8 sequence,
                //  treat it as an identifier
                if !invalid_sequence_started && !c.is_ascii() && c.is_alphabetic() {
                    return Some(self.produce_non_ascii_identifier(pos));
                }

                if !invalid_sequence_started {
                    invalid_sequence_started = true;
                    self.errors.log_error(error::new_invalid_sequence_error(
                        pos,
                        self.reader.pos() - pos,
                    ));
                } else {
                    // TODO: Check that last error len matches current posision
                    //  there might be spacing in-between
                    self.errors.adjust_last_error_end(self.reader.pos());
                }
            }
        }

        return None;
    }
}

impl<'a, R: Read + Seek, S: source::Source<'a, R>> ScannerImpl<'a, R, S> {
    pub fn new(source: &'a S) -> Self {
        ScannerImpl {
            source: source,
            reader: source::LookAheadReader::new(source.get_readable()),
            allow_indentation: true,
            errors: error::ErrorManager::new(),
        }
    }

    fn get_source_string(&self, pos: u64, len: usize) -> String {
        let mut reader = self.source.get_readable();
        match reader.seek(SeekFrom::Start(pos)) {
            Ok(_) => {
                let mut v: Vec<u8> = Vec::new();
                v.resize(len, 0);
                if let Ok(_) = reader.read(&mut v) {
                    return String::from_utf8_lossy(&v).to_string();
                }
            }
            Err(_) => (),
        }
        return "".into();
    }

    fn read_utf8_char(&mut self) -> Option<char> {
        let controlbyte = self.reader.peek().unwrap();

        self.reader.advance();

        if controlbyte <= 127 {
            return Some(controlbyte as char);
        }

        // Utf-8 is at max 4 bytes
        let mut buf: [u8; 4] = [controlbyte, 0, 0, 0];

        const CTRL_2_BYTE_MASK: u8 = 0b11100000;
        const CTRL_3_BYTE_MASK: u8 = 0b11110000;
        const CTRL_4_BYTE_MASK: u8 = 0b11111000;
        const CTRL_2_BYTE_VALUE: u8 = 0b11000000;
        const CTRL_3_BYTE_VALUE: u8 = 0b11100000;
        const CTRL_4_BYTE_VALUE: u8 = 0b11110000;

        // Check how long the sequence should be
        let bytecount;
        match controlbyte {
            n if n & CTRL_2_BYTE_MASK == CTRL_2_BYTE_VALUE => bytecount = 2,
            n if n & CTRL_3_BYTE_MASK == CTRL_3_BYTE_VALUE => bytecount = 3,
            n if n & CTRL_4_BYTE_MASK == CTRL_4_BYTE_VALUE => bytecount = 4,
            _ => return None,
        }

        for i in 1..bytecount {
            let n = self.reader.peek();
            self.reader.advance();
            match n {
                Some(n) => buf[i] = n,
                _ => return None,
            }
        }

        let res = std::str::from_utf8(&buf[0..bytecount]);
        if res.is_ok() {
            return res.unwrap().chars().nth(0);
        }

        return None;
    }

    fn read_utf8_char_with_error(&mut self) -> Option<char> {
        let pos = self.reader.pos();
        return match self.read_utf8_char() {
            Some(n) => Some(n),
            None => {
                self.errors.log_error(error::new_non_utf8_sequence_error(
                    pos,
                    self.reader.pos() - pos,
                ));
                None
            }
        };
    }

    fn produce_linecomment(&mut self) -> Token {
        let startpos = self.reader.pos();
        debug_assert!(self.reader.peek().unwrap() == b'/');
        self.reader.advance();
        debug_assert!(self.reader.peek().unwrap() == b'/');
        self.reader.advance();

        // Eat until line break
        while let Some(n) = self.reader.peek() {
            if n == b'\n' {
                break;
            }
            self.reader.advance();
        }

        return Token::new(
            TokenType::Comment,
            startpos,
            (self.reader.pos() - startpos) as usize,
        );
    }

    fn produce_blockcomment(&mut self) -> Token {
        let startpos = self.reader.pos();
        debug_assert!(self.reader.peek().unwrap() == b'/');
        self.reader.advance();
        debug_assert!(self.reader.peek().unwrap() == b'*');
        self.reader.advance();

        // Eat block, including nested blocks
        let mut blocklevel = 1;
        while let Some(n) = self.reader.peek() {
            if n == b'/' && self.reader.lookahead() == Some('*' as u8) {
                blocklevel += 1;
                self.reader.advance();
            } else if n == b'*' && self.reader.lookahead() == Some('/' as u8) {
                blocklevel -= 1;
                self.reader.advance();
            }

            self.reader.advance();

            if blocklevel == 0 {
                break;
            }
        }

        if blocklevel != 0 {
            // TODO: Add error reference to start of comment
            self.errors.log_error(error::new_unexpected_eof_error(
                self.reader.pos(),
                "Unexpected end of file inside block comment".into(),
            ));
        }

        return Token::new(
            TokenType::Comment,
            startpos,
            (self.reader.pos() - startpos) as usize,
        );
    }

    fn consume_spacing(&mut self) {
        debug_assert!(self.reader.peek().unwrap() == b' ');
        self.reader.advance();

        while let Some(n) = self.reader.peek() {
            if n != b' ' {
                break;
            }
            self.reader.advance();
        }
    }

    fn produce_indentation(&mut self) -> Token {
        let startpos = self.reader.pos();
        debug_assert!(self.reader.peek().unwrap() == b'\t');
        self.reader.advance();

        while let Some(n) = self.reader.peek() {
            if n != b'\t' {
                break;
            }
            self.reader.advance();
        }
        return Token::new(
            TokenType::Indentation,
            startpos,
            (self.reader.pos() - startpos) as usize,
        );
    }

    // Produces an "invalid" identifier and logs an error
    fn produce_non_ascii_identifier(&mut self, sourcepos: u64) -> Token {
        // Error recovery: advance until end of non-ascii utf8 sequence
        while let Some(n) = self.reader.peek() {
            if (n as char).is_ascii() && !(n as char).is_ascii_alphanumeric() {
                break;
            }

            self.read_utf8_char_with_error();
        }

        self.errors.log_error(error::new_non_ascii_identifier_error(
            sourcepos,
            self.reader.pos() - sourcepos,
            self.get_source_string(sourcepos, (self.reader.pos() - sourcepos) as usize),
        ));

        return Token::new(
            TokenType::Identifier,
            sourcepos,
            (self.reader.pos() - sourcepos) as usize,
        );
    }

    // Produce identifier starting at supplied source pos, continuing at the reader pos
    fn produce_identifier_at_pos(&mut self, sourcepos: u64) -> Token {
        // TODO: This seems unnecessary
        let mut string: String = String::new();

        while let Some(n) = self.reader.peek() {
            if !(n as char).is_ascii_alphanumeric() && n != b'_' {
                // If we are still ascii, this marks the end of a valid identifier
                if (n as char).is_ascii() {
                    break;
                }

                // In order to not give cascading errors in the parser, we still produce a token here
                return self.produce_non_ascii_identifier(sourcepos);
            }
            string.push(n as char);
            self.reader.advance();
        }

        // Everything was fine, check if this was a keyword
        if let Some(tokentype) = KEYWORDS.get(&string) {
            return Token::new(
                *tokentype,
                sourcepos,
                (self.reader.pos() - sourcepos) as usize,
            );
        }

        // If not a keyword, it is an identifier
        return Token::new(
            TokenType::Identifier,
            sourcepos,
            (self.reader.pos() - sourcepos) as usize,
        );
    }

    // Produce identifier at reader pos
    fn produce_identifier(&mut self) -> Token {
        return self.produce_identifier_at_pos(self.reader.pos());
    }

    fn produce_stringliteral(&mut self) -> Token {
        let startpos = self.reader.pos();
        debug_assert!(self.reader.peek().unwrap() == b'\"');
        self.reader.advance();

        while self.reader.peek().filter(|c| *c != b'\"').is_some() {
            self.reader.advance();
        }

        // TODO: Add error reference to start of literal
        if self.reader.peek().is_none() {
            self.errors.log_error(error::new_unexpected_eof_error(
                self.reader.pos(),
                "Unexpected end of file inside string literal".into(),
            ));
        } else {
            self.reader.advance();
        }

        return Token::new(
            TokenType::StringLiteral,
            startpos,
            (self.reader.pos() - startpos) as usize,
        );
    }

    fn produce_characterliteral(&mut self) -> Token {
        let startpos = self.reader.pos();
        debug_assert!(self.reader.peek().unwrap() == b'\'');
        self.reader.advance();

        while self.reader.peek().filter(|c| *c != b'\'').is_some() {
            self.reader.advance();
        }

        // TODO: Add error reference to start of literal
        if self.reader.peek().is_none() {
            self.errors.log_error(error::new_unexpected_eof_error(
                self.reader.pos(),
                "Unexpected end of file inside character literal".into(),
            ));
        } else {
            self.reader.advance();
        }

        return Token::new(
            TokenType::CharacterLiteral,
            startpos,
            (self.reader.pos() - startpos) as usize,
        );
    }

    fn produce_numericliteral(&mut self) -> Token {
        let startpos = self.reader.pos();
        debug_assert!(self.reader.peek().unwrap().is_ascii_digit());
        self.reader.advance();

        // Note: we eat all trailing alphanumericals in this function, parsing of the
        //  actual number and error reporting happens when constructing AST

        while self
            .reader
            .peek()
            .filter(|c| c.is_ascii_alphanumeric() || *c == b'.')
            .is_some()
        {
            self.reader.advance();
        }

        return Token::new(
            TokenType::NumericLiteral,
            startpos,
            (self.reader.pos() - startpos) as usize,
        );
    }

    fn produce_token_and_advance(&mut self, tokentype: TokenType) -> Token {
        let token = Token::new(tokentype, self.reader.pos(), 1);
        self.reader.advance();
        return token;
    }

    fn produce_token_and_advance_n(&mut self, tokentype: TokenType, len: usize) -> Token {
        let token = Token::new(tokentype, self.reader.pos(), len);
        for _ in 0..len {
            self.reader.advance();
        }
        return token;
    }
}
