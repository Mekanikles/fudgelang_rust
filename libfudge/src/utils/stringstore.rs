use crate::utils::*;

pub type StringStore = objectstore::HashedObjectStore<StringKey, String>;

impl objectstore::HashedStoreKey<String> for StringKey {
    fn from_obj(object: &String) -> Self {
        StringKey::from_str(object.as_str())
    }
}
