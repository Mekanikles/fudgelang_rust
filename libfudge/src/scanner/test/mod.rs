use super::scanner::*;
use super::token::*;

mod utils;
use utils::*;

mod arithmetics;
mod basic;
mod brackets;
mod characterliterals;
mod comments;
mod identifiers;
mod indentation;
mod keywords;
mod misctokens;
mod numericliterals;
mod stringliterals;

#[test]
fn test_file_with_comments() {
    let scanner_result = get_scanner_result_from_file("testdata/comments.fu");
    // Needs to be called here directly to retain snaphot file name
    insta::assert_debug_snapshot!(&scanner_result.tokens);
}
