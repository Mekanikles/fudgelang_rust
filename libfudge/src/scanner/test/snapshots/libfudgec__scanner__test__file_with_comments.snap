---
source: libfudge/src/scanner/test/mod.rs
assertion_line: 23
expression: "&scanner_result.tokens"
---
[
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 0,
            len: 3,
        },
    },
    Token {
        tokentype: OpeningParenthesis,
        source_span: SourceSpan {
            pos: 3,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 4,
            len: 5,
        },
    },
    Token {
        tokentype: Comma,
        source_span: SourceSpan {
            pos: 9,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 11,
            len: 7,
        },
    },
    Token {
        tokentype: ClosingParenthesis,
        source_span: SourceSpan {
            pos: 18,
            len: 1,
        },
    },
    Token {
        tokentype: Comment,
        source_span: SourceSpan {
            pos: 20,
            len: 10,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 30,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 31,
            len: 3,
        },
    },
    Token {
        tokentype: OpeningParenthesis,
        source_span: SourceSpan {
            pos: 34,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 35,
            len: 5,
        },
    },
    Token {
        tokentype: Comma,
        source_span: SourceSpan {
            pos: 40,
            len: 1,
        },
    },
    Token {
        tokentype: Comment,
        source_span: SourceSpan {
            pos: 41,
            len: 18,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 60,
            len: 5,
        },
    },
    Token {
        tokentype: ClosingParenthesis,
        source_span: SourceSpan {
            pos: 65,
            len: 1,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 66,
            len: 1,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 67,
            len: 1,
        },
    },
    Token {
        tokentype: Comment,
        source_span: SourceSpan {
            pos: 68,
            len: 275,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 343,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 344,
            len: 3,
        },
    },
    Token {
        tokentype: OpeningParenthesis,
        source_span: SourceSpan {
            pos: 347,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 348,
            len: 7,
        },
    },
    Token {
        tokentype: Comma,
        source_span: SourceSpan {
            pos: 355,
            len: 1,
        },
    },
    Token {
        tokentype: Identifier,
        source_span: SourceSpan {
            pos: 357,
            len: 5,
        },
    },
    Token {
        tokentype: ClosingParenthesis,
        source_span: SourceSpan {
            pos: 362,
            len: 1,
        },
    },
    Token {
        tokentype: LineBreak,
        source_span: SourceSpan {
            pos: 363,
            len: 1,
        },
    },
    Token {
        tokentype: Comment,
        source_span: SourceSpan {
            pos: 364,
            len: 16,
        },
    },
]
